### ISSUE TITLE
r-theta runs cannot be restarted

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#41]
If one runs a model in an rtheta grid, then an attempt to restart the model to add more cycles results in a segmentation falult.
This is generated by rtheta_ds_in_cell, when it attempts to read the cone structure.
I think the reason for this is that the cone structure is not saved as part of the wind.save file. For a cylindrical grid, python.c recomputes the cones upon restart, but I don't think the rtheta cones are regenerated, hance the structure is empty.

Two possible solutions:

Write out the cones as part of the wind.save
Regenerate the cones on restart.

**Comment by Higginbottom** [comment]
This issue has been reopened - changes made in Jan 18 meant that the above fix stopped working. It was a simple problem - the above fix relied upon a check 

 if (zdom[ndom].coord_type == RTHETA && geo.run_type == SYSTEM_TYPE_PREVIOUS)

to regenerate the cones_rtheta via my little subroutine. Changes meant that a restart is now defined as RUN_TYPE_RESTART (1) whereas SYSTEM_TYPE_PREVIOUS is 3. This meant that the little routine is no longer called, and so the code crashes the first time it goes to theta_ds_in_cell because cones_rtheta is not defined. 

A simple fix is obvious, but I wanted to quickly try and see if there is something more pernicious going on.

**Comment by Higginbottom** [comment]
Question: Why does this problem only affect an r-theta run and not a cylindrical run. It turns out that there are (at least) two cones defined in python.

There is one type of cone defined in the domains which is used to decide if a photon has entered the wind - it is used in ds_to_cone.

ds_to_cone (&zdom[ndom].windcone[0]

this works fine. These wind cones are set up using setup_windcone - called *only* in python.c

Another type of wind cone is zdom[ndom].cones_rtheta which is a special type of wind cone used for r_theta. I *think* this is used as a shortcut to see if photons leave cells in the theta direction. This is used in 

ds_in_cell

which has a test in it to see what coord type we are using, if rtheta it subsequently calls

theta_ds_in_cell which has a call to 

ds_to_cone (&zdom[ndom].cones_rtheta[iz], p);

this uses the machinery of the cones to see if a photon leaves a cell in the theta direction. It is not used in cylindrical cells for obvious reasons.

So, this is why the problem only raises its head in r-theta - it is not *really* a cone problem.

Solution - just fix the call to use RUN_TYPE_RESTART in my little test so we call rtheta_make_cones in the event of a restart. It should probably also be called in the case of a previous run as well. I'll just do a little test...

**Comment by Higginbottom** [comment]
OK - one does also need to do it for a restart.

But thinking more generally, I wonder if these cones_rtheta should be a part of the domain - they seem to really be a part of the wind. If they were in the wind, then they would get saved as part of the wind save. This confusion is because we are essentially re-using part of the code that is intended to see if we get into the wind (or a domain) to see if we move from one cell to another in the wind....

**Comment by kslong** [comment]
Some of this is too terse for me to understand. For example the comments about code changes don't say what routine we are discussing.  But I do think there are issues regarding the hydro model and domains, not so much because I expect we will use mulitiple domains with zeus, but because they are treated as a special case.  don't fit into the way we are doing things more generally.  

With regard to cones, and other boundaries  There should be two issues to worry about, one is the overall boundary of all of the winds, and the other is the boundaries for each of the wind domains.  Without having looked at this in detail, the boundaries associated with the domains should be part of the domain structure, and the ones that have to do with the overall wind should be associated with geo.   The domain prt does have cones.  

Note that we are having a lot of problems with this right now with importing models more generally, so it would be good to cleanly sort all of this stuff out.

**Comment by Higginbottom** [comment]
So the problem we are having is not the boundaries associated with the domains, or the overall wind, but cells within the wind where the 'cones' structures are re-puposed to work out if photons are leaving cells in an theta-direction. This is in the subroutine rtheta_ds_in_cell which is in the file rtheta.c. This is nothing in particular to do with hydro, but is to do with r-theta geometry in general. I have reproduced the problem with a modified cv.pf file - I just use the cv.pf file in examples and set the geom type to r-theta. Run a cycle, then stop and restart for an extra cycle. It will crash.
I'm pulling a fix to my fix, but as you say we may wish to fix this all long term with a big rethink.

**Comment by kslong** [comment]
That does sound like a different problem, and something I missed when I was trying to make restarts and previous work in a more consistent fashion.  Mea culpa.

**Comment by Higginbottom** [comment]
fixed with pull #357

**Comment by kslong** [comment]
It is not entirely clear that this was the best solution for this problem, or rather that one needed to have special logic to handle this case.  rtheta_make_cones should probably have been called from rtheta_wind_complete as this would ensure that whenever a windsave file is read in, that the the cones would be populated.  Alternatively as @Higginbottom  had noted we could have written out the cones (which are in the domain structure).   I guess the reason that we did not do that, is that the cones are dynamically allocated, unlike the midpts.  

I am reopening this briefly, while we decide if there is anything we want to do.

**Comment by Higginbottom** [comment]
I don't disagree with any of that - my fix was a fix to a fix that was applied a long time ago. You are right that it would make more sense to put this as part of rtheta_wind_complete - assuming that this routine is called whenever we restart an r-theta wind or use a previous model in r-theta....

**Comment by Higginbottom** [comment]
A quick test seems to confirm that one can just add 

  rtheta_make_cones (ndom, w);

to the end of rtheta_wind_complete to replace the if statement in python.c. This is certainly cleaner...

**Comment by Higginbottom** [comment]
OK, so I wanted to do a bit of archaeology and see where all this started.
It began back in the mists of time when I started wanting to import r-theta models for the first paper we did with Proga.
The code to import the Proga models was broken, and I rebuilt it. 
In r-theta coordinates, the cone architecture is used to track photons in the theta direction, and the cones used for this were originally computed as a part of rtheta_make_grid subroutine in rtheta.c. However, when importing a model, one wants to set up the grid to match the input geometry (or at least that was the decision I made) so this routine isn't called. I therefore broke out the commands to make the cones into a separate routine that was then called in rtheta_make_grid and also in the hydro import routine.
This also allowed me to fix bug #41. The root cause of this bug is that rtheta_make_grid isn't called in a restart (or previous) situation, and so the cones aren't made and so the first time one tries to move a photon in the theta direction (using rtheta_ds_in_cell) one gets a seg fault.
I chose to do this in python.c with a call to my new routine in the case of an rtheta restart. This made sense at the time there were many lines of code in python.c defining cones and I was mistaken in thinking that this cone work was similar - in fact (as discussed above) it is a very different thing that is best done when the wind is completed. 
When all the cones stuff was removed from python.c this bit of code got marooned - doubtless because KSL was (understandably) confused by it - as evidenced by some XXX comments!
